pkg/controller/cyclenoderequest/transitioner/transitions_successful_test.go

```diff
@@ -0,0 +1,275 @@
+}
+
+// Test that if the phase executes after the CNR deletion expiry time then the
+// CNR will be deleted.
+func TestSuccessfulBeforeDeleteTime(t *testing.T) {
+	cnr := &v1.CycleNodeRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              "cnr-1",
+			Namespace:         "kube-system",
+			CreationTimestamp: metav1.Now(),
+		},
+		Status: v1.CycleNodeRequestStatus{
+			Phase:   v1.CycleNodeRequestSuccessful,
+			Message: "",
+		},
+	}
+
+	fakeTransitioner := NewFakeTransitioner(cnr,
+		WithTransitionerOptions(Options{
+			DeleteCNR:       true,
+			DeleteCNRExpiry: 5 * time.Second,
+		}),
+	)
+
+	var list v1.CycleNodeRequestList
+
+	assert.NoError(t,
+		fakeTransitioner.Client.K8sClient.List(context.TODO(), &list, &client.ListOptions{}),
+	)
+
+	assert.Len(t, list.Items, 1)
+
+	// Execute the Successful phase
+	_, err := fakeTransitioner.Run()
+	assert.NoError(t, err)
+
+	assert.NoError(t,
+		fakeTransitioner.Client.K8sClient.List(context.TODO(), &list, &client.ListOptions{}),
+	)
+
+	assert.Len(t, list.Items, 1)
+}
+
+// Test that the Successful phase will delete sibling CNRs for the same
+// nodegroup that are in the failed phase. No other CNRs should be deleted.
+func TestSuccessfulDeleteFailedSiblingCNRs(t *testing.T) {
+	// CNR used for the execution
+	// Should not be deleted
+	cnr1 := &v1.CycleNodeRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              "cnr-1",
+			Namespace:         "kube-system",
+			CreationTimestamp: metav1.Now(),
+		},
+		Spec: v1.CycleNodeRequestSpec{
+			NodeGroupName: "ng-1",
+		},
+		Status: v1.CycleNodeRequestStatus{
+			Phase:   v1.CycleNodeRequestSuccessful,
+			Message: "",
+		},
+	}
+
+	// Failed CNR for the same nodegroup
+	// Should be deleted
+	cnr2 := &v1.CycleNodeRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              "cnr-2",
+			Namespace:         "kube-system",
+			CreationTimestamp: metav1.Now(),
+		},
+		Spec: v1.CycleNodeRequestSpec{
+			NodeGroupName: "ng-1",
+		},
+		Status: v1.CycleNodeRequestStatus{
+			Phase:   v1.CycleNodeRequestFailed,
+			Message: "",
+		},
+	}
+
+	// Pending CNR for the same nodegroup
+	// Should NOT be deleted
+	cnr3 := &v1.CycleNodeRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              "cnr-3",
+			Namespace:         "kube-system",
+			CreationTimestamp: metav1.Now(),
+		},
+		Spec: v1.CycleNodeRequestSpec{
+			NodeGroupName: "ng-1",
+		},
+		Status: v1.CycleNodeRequestStatus{
+			Phase:   v1.CycleNodeRequestPending,
+			Message: "",
+		},
+	}
+
+	// Failed CNR for a different nodegroup
+	// Should NOT be deleted
+	cnr4 := &v1.CycleNodeRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              "cnr-4",
+			Namespace:         "kube-system",
+			CreationTimestamp: metav1.Now(),
+		},
+		Spec: v1.CycleNodeRequestSpec{
+			NodeGroupName: "ng-2",
+		},
+		Status: v1.CycleNodeRequestStatus{
+			Phase:   v1.CycleNodeRequestFailed,
+			Message: "",
+		},
+	}
+
+	// Failed CNR for the same nodegroup in a different namespace
```

Comment:
Are CNRs intended to be namespaced? I can't see much value in separating them by namespace like this but being able to act on the same set of backing nodes provided by a cloud provider.

Reply:
That would go back to when they were designed. I'm not particularly sure, though changing that is out of scope here.